### ISSUE TITLE
Explode WAR instead of working on exploded WAR directory for Maven

### ISSUE THREAD

**Issue by chanseokoh** [GoogleContainerTools/jib#1091]
Explode the WAR file rather than using this directory. The contents of the final WAR may be different from this directory (it's possible to include or exclude files when packaging a WAR. Also the exploded WAR directory is configurable with `<webappDirectory>` and may not be at `build.getFinalName()`.

https://github.com/GoogleContainerTools/jib/pull/1068/files#r221696396

**Comment by chanseokoh** [comment]
FYI @edrandall

**Comment by edrandall** [comment]
It's been a long-standing convention in all of our web applications (we have a large number) that we store versioning information about the build within the MANIFEST.MF.  We have an API which makes this available to monitoring and healthcheck systems.  We absolutely need this information present in the build as we migrate to cloud.

META-INF/MANIFEST.MF is part of the .jar file spec and as such is pretty much guaranteed to be present, it's designed for storing exactly this type information as well as digital signatures.

Just curious, where is it documented that Jib actually changes the deployed files in this way, a brief search found nothing (not even this ticket)?

**Comment by chanseokoh** [comment]
Thanks for the detailed explanation. That makes sense.

Jib doesn't change the deployed files. It is just that by default, Jib is supposed to put exploded WAR contents, but because of this issue that it is using the intermediate exploded WAR directory prepared by Maven (whose contents are often same as the final WAR) instead of actually exploding the final WAR, incomplete `MANIFEST/MANIFEST.MF` will be put into the image. I think once we fix this, your issue will be resolved.

BTW, Jib Gradle does explode the final WAR, so this issue is present only for Maven.

**Comment by edrandall** [comment]
Thanks - so it's something of a side-effect of the order in which Maven builds things, have I got that right?
Any idea how soon this can be fixed?  We're trying to migrate a lot of projects and Jib is a good strategic solution, but  we may need to implement a tactical workaround depending on timeline.

**Comment by chanseokoh** [comment]
Yeah, Maven prepares exploded WAR contents in a build output directory before finally zipping them up into a WAR file, and we've been lazy to use that intermediate directory instead of unzipping the WAR ourselves. This works in many cases, as the contents are often same except for files like `MANIFEST.MF`, but yeah, it is time to fix this.

This should be easy to fix, so I will see if we can fix this before our next release.

In the meantime, if you have a Gradle project, you may try Jib Gradle just to see if it works as desired.

**Comment by edrandall** [comment]
The following maven 'hack' seems to provide a crude workaround:
```
			<plugin>
				<groupId>org.apache.maven.plugins</groupId>
				<artifactId>maven-dependency-plugin</artifactId>
				<version>3.1.1</version>
				<executions>
					<execution>
						<id>unpack-meta-inf</id>
						<phase>package</phase>
						<goals>
							<goal>unpack</goal>
						</goals>
						<configuration>
							<artifactItems>
								<artifactItem>
									<groupId>${project.groupId}</groupId>
									<artifactId>${project.artifactId}</artifactId>
									<version>${project.version}</version>
									<type>war</type>
									<overWrite>true</overWrite>
									<outputDirectory>${project.build.directory}/${project.build.finalName}</outputDirectory>
									<includes>META-INF/**</includes>
								</artifactItem>
							</artifactItems>
						</configuration>
					</execution>
				</executions>
			</plugin>
```

**Comment by TadCordle** [comment]
@edrandall We just released 1.7.0 with this fix!

**Comment by TadCordle** [comment]
@edrandall @joelhandwell @Daiki-Kawanuma @cessien @kurczynski @CauchyPeano We've released v1.7.0 with this fix!
